### PR TITLE
Small improvement of user experience

### PR DESCRIPTION
**Pull request by razvanstoica89**
Add details about remediation measures for host configuration tests

### CODE REVIEW

**Comment by konstruktoid** [comment]
Thanks @razvanstoica89, and it was a humble title you picked, calling this PR "small".

File: docker-bench-security.sh [hunk 1]
```diff
@@ -2,7 +2,7 @@
 # ------------------------------------------------------------------------------
 # Docker Bench for Security
 #
-# Docker, Inc. (c) 2015-
+# Docker, Inc. (c) 2015-2021
```
[the remark this answers is not in the document]
Comment (razvanstoica89) [reply docker-bench-security.sh:5]: Unfortunately I can't because that code is a comment. It will never be executed in order to be able to calculate the current date as in case #466.

[the remark this answers is not in the document]
Comment (konstruktoid) [reply docker-bench-security.sh:5]: yeah, silly me. I read it as the `yell` part.

File: tests/1_host_configuration.sh [hunk 2]
```diff
@@ -59,18 +59,19 @@ check_1_1_2() {
...
 
 # 1.2.1
 check_1_2_1() {
-  id_1_2_1="1.2.1"
-  desc_1_2_1="Ensure a separate partition for containers has been created (Scored)"
-  check_1_2_1="$id_1_2_1 - $desc_1_2_1"
-  starttestjson "$id_1_2_1" "$desc_1_2_1"
+  local id="1.2.1"
+  local desc="Ensure a separate partition for containers has been created (Scored)"
+  local remediation="For new installations, you should create a separate partition for the /var/lib/docker mount point. For systems that have already been installed, you should use the Logical Volume Manager (LVM) within Linux to create a new partition."
```
Comment (konstruktoid) [inline tests/1_host_configuration.sh:72]: I really like the `remediation` stuff, but it gets very very spammy. Collect the all and present it as a list in the end?


```sh
[INFO] 1.2 - Linux Hosts Specific Configuration
[WARN] 1.2.1 - Ensure a separate partition for containers has been created (Scored)
[INFO]        * Remediation: For new installations, you should create a separate partition for the /var/lib/docker mount point. For systems that have already been installed, you should use the Logical Vol
ume Manager (LVM) within Linux to create a new partition.
[INFO] 1.2.2  - Ensure only trusted users are allowed to control Docker daemon (Scored)
[INFO]        * docker:x:997:vagrant   
[WARN] 1.2.3  - Ensure auditing is configured for the Docker daemon (Scored)
[INFO]        * Remediation: Install and configure auditd using the following command: sudo apt-get install auditd. Add the following line to the /etc/audit/rules.d/audit.rules file: -w /usr/bin/dockerd -
k docker. Then restart the audit daemon using the following command: service auditd restart.
[INFO]        * Remediation impact: Because the audit can generate large log files, you need to make sure that they are rotated and archived periodically. It is a good idea to create a separate partition 
for audit logs to avoid filling up other critical partitions.
[WARN] 1.2.4  - Ensure auditing is configured for Docker files and directories - /var/lib/docker (Scored)
[INFO]        * Remediation: Install and configure auditd using the following command: sudo apt-get install auditd. Add the following line to the /etc/audit/rules.d/audit.rules file: -w /var/lib/docker -k
 docker. Then restart the audit daemon using the following command: service auditd restart.
[INFO]        * Remediation impact: Because the audit can generate large log files, you need to make sure that they are rotated and archived periodically. It is a good idea to create a separate partition 
for audit logs to avoid filling up other critical partitions.
[WARN] 1.2.5  - Ensure auditing is configured for Docker files and directories - /etc/docker (Scored)
[INFO]        * Remediation: Install and configure auditd using the following command: sudo apt-get install auditd. Add the following line to the /etc/audit/rules.d/audit.rules file: -w /etc/docker -k doc
ker. Then restart the audit daemon using the following command: service auditd restart.
[INFO]        * Remediation impact: Because the audit can generate large log files, you need to make sure that they are rotated and archived periodically. It is a good idea to create a separate partition 
for audit logs to avoid filling up other critical partitions.
[WARN] 1.2.6  - Ensure auditing is configured for Docker files and directories - docker.service (Scored)
[INFO]        * Remediation: Install and configure auditd using the following command: sudo apt-get install auditd. Add the following line to the /etc/audit/rules.d/audit.rules file: -w /lib/systemd/syste
m/docker.service -k docker. Then restart the audit daemon using the following command: service auditd restart.
[INFO]        * Remediation impact: Because the audit can generate large log files, you need to make sure that they are rotated and archived periodically. It is a good idea to create a separate partition 
for audit logs to avoid filling up other critical partitions.
[WARN] 1.2.7  - Ensure auditing is configured for Docker files and directories - docker.socket (Scored)
[INFO]        * Remediation: Install and configure auditd using the following command: sudo apt-get install auditd. Add the following line to the /etc/audit/rules.d/audit.rules file: -w /lib/systemd/syste
m/docker.socket -k docker. Then restart the audit daemon using the following command: service auditd restart.
[INFO]        * Remediation impact: Because the audit can generate large log files, you need to make sure that they are rotated and archived periodically. It is a good idea to create a separate partition 
for audit logs to avoid filling up other critical partitions.
[WARN] 1.2.8  - Ensure auditing is configured for Docker files and directories - /etc/default/docker (Scored)
[INFO]        * Remediation: Install and configure auditd using the following command: sudo apt-get install auditd. Add the following line to the /etc/audit/rules.d/audit.rules file: -w /etc/default/docke
r -k docker. Then restart the audit daemon using the following command: service auditd restart.
[INFO]        * Remediation impact: Because the audit can generate large log files, you need to make sure that they are rotated and archived periodically. It is a good idea to create a separate partition 
for audit logs to avoid filling up other critical partitions.
[INFO] 1.2.9  - Ensure auditing is configured for Docker files and directories - /etc/sysconfig/docker (Scored)
[INFO]        * File not found
```

Comment (razvanstoica89) [reply tests/1_host_configuration.sh:72]: I fully agree with you. I think it's a good idea to collect them and print them all at the end.

File: tests/5_container_runtime.sh [hunk 3]
```diff
@@ -1047,30 +1047,30 @@ check_5_26() {
...
+  local id="5.26"
+  local desc="Ensure that container health is checked at runtime (Scored)"
+  local check="$id  - $desc"
+  starttestjson "$id" "$desc"
 
   totalChecks=$((totalChecks + 1))
 
   fail=0
-  nohealthcheck_containers=""
+  nohealthlocal check=""
```
[the remark this answers is not in the document]
Comment (razvanstoica89) [reply tests/5_container_runtime.sh:1058]: It is my mistake. I shouldn't have changed anything there.

**Comment by konstruktoid** [comment]
Loads of updates 👍 
Will test as soon as possible.

**Comment by konstruktoid** [comment]
Can you poke me @razvanstoica89 when you believe you are finished with this PR so I can start testing?
No need to test if it's a work in progress.

**Comment by razvanstoica89** [comment]
Yes of course. Thanks

File: docker-bench-security.sh [hunk 4]
```diff
@@ -78,7 +95,7 @@ do
 done
 
 if [ -z "$logger" ]; then
-  logger="${myname}.log"
+  logger="log/${myname}.log"
```
Comment (konstruktoid) [inline docker-bench-security.sh:98]: make sure that directory exists

File: distros/Dockerfile.alpine [hunk 5]
```diff
@@ -1,25 +0,0 @@
-FROM alpine:3.12
```
Comment (konstruktoid) [inline distros/Dockerfile.alpine:1]: Why is this file deleted, and not replaced with the main directory Dockerfile?

Comment (razvanstoica89) [reply distros/Dockerfile.alpine:1]: Because it is a duplicate file. The same file is in the main directory and the distros directory. If you don't like this, I'll put the Dockerfile from the main directory also in distros directory.

**Comment by razvanstoica89** [comment]
@konstruktoid Let me know if you want me to make any changes or if you find any bugs. In the coming days i will add more remediation stuff. After that I will finish this pull-request.

File: distros/Dockerfile.alpine (hunk 5 above)
Comment (konstruktoid) [reply distros/Dockerfile.alpine:1]: Nah, delete it. I'll go over the Dockerfiles in the `distros` directory later.

**Comment by razvanstoica89** [comment]
@konstruktoid I think I'm done with this PR.

**Comment by konstruktoid** [comment]
Will start testing as soon as possible.

File: docker-bench-security.sh [hunk 6]
```diff
@@ -103,11 +132,15 @@ fi
...
 currentScore=0
 
-logit "Initializing $(date)\n"
+logit "Initializing $(date +%Y-%m-%dT%H:%M:%S%:z)\n"
+appendjson
 beginjson "$version" "$(date +%s)"
 
 # Load all the tests from tests/ and run them
 main () {
+  logit "\n${bldylw}Section A - Checks result${txtrsr}"
```
Comment (konstruktoid) [inline docker-bench-security.sh:141]: Nitpicking - "Check results"? Dont know which one is more correct english.

File: tests/1_host_configuration.sh [hunk 7]
```diff
@@ -2,474 +2,443 @@
...
 
-# 1.2.3
 check_1_2_3() {
-  id_1_2_3="1.2.3"
-  desc_1_2_3="Ensure auditing is configured for the Docker daemon (Scored)"
-  check_1_2_3="$id_1_2_3  - $desc_1_2_3"
-  starttestjson "$id_1_2_3" "$desc_1_2_3"
+  local id="1.2.3"
+  local desc="Ensure auditing is configured for the Docker daemon (Scored)"
+  local remediation="Install and configure auditd using command sudo apt-get install auditd. Add -w /usr/bin/dockerd -k docker to the /etc/audit/rules.d/audit.rules file. Then restart the audit daemon using command service auditd restart."
```
Comment (konstruktoid) [inline tests/1_host_configuration.sh:132]: The package manager stuff needs to be generic, `apt-get` is Debian family for example

**Comment by konstruktoid** [comment]
the json file doesnt seem to be valid.
```sh
$  jq log/docker-bench-security.sh.log.json 
jq: error: docker/0 is not defined at <top-level>, line 1:
log/docker-bench-security.sh.log.json    
jq: error: bench/0 is not defined at <top-level>, line 1:
log/docker-bench-security.sh.log.json           
jq: error: security/0 is not defined at <top-level>, line 1:
log/docker-bench-security.sh.log.json                 
jq: 3 compile errors
```

**Comment by konstruktoid** [comment]
Since you've created the `log` directory, perhaps rename without the `.sh` extension

File: docker-bench-security.sh (hunk 6 above)
Comment (razvanstoica89) [reply docker-bench-security.sh]: It's done

**Comment by razvanstoica89** [comment]
> the json file doesnt seem to be valid.
> 
> ```shell
> $ jq log/docker-bench-security.sh.log.json 
> jq: error: docker/0 is not defined at <top-level>, line 1:
> log/docker-bench-security.sh.log.json    
> jq: error: bench/0 is not defined at <top-level>, line 1:
> log/docker-bench-security.sh.log.json           
> jq: error: security/0 is not defined at <top-level>, line 1:
> log/docker-bench-security.sh.log.json                 
> jq: 3 compile errors
> ```

Looks like you're trying to validate that the string "log/docker-bench-security.sh.log.json" is a valid json.
I think a better approach is `cat log/docker-bench-security.log.json |  jq empty`.

As far as I know the only way you can get an invalid json log is to stop the script manually while it is running. Even if you run it again, you will be left with an invalid json log.

**Comment by konstruktoid** [comment]
Running a standrd installation without any containers:

```sh
Section A - Check results

[INFO] 1 - Host Configuration
[INFO] 1.1 - General Configuration
[NOTE] 1.1.1  - Ensure the container host has been Hardened (Not Scored)
[INFO] -c
[INFO]        * Using 20.10.5, verify is it up to date as deemed necessary
[INFO]        * Your operating system vendor may provide support and security maintenance for Docker
[INFO] 1.2 - Linux Hosts Specific Configuration
WARNING: No swap limit support
[WARN] -s
[INFO] -c
[INFO]       * Users: vagrant
[WARN] -s
[WARN] -s
[WARN] -s
[WARN] -s
[WARN] -s
[WARN] -s
[INFO] -c
[INFO]        * File not found
[INFO] -c
[INFO]         * File not found
[WARN] -s
[INFO] -c
[INFO]         * File not found

[INFO] 2 - Docker daemon configuration 
[WARN] -s
[PASS] -s
[PASS] -s
[PASS] -s
[PASS] -s
[INFO] -c
[INFO]      * Docker daemon not listening on TCP
[INFO] -c
[INFO]      * Default ulimit doesn't appear to be set
[WARN] -s
[PASS] -s
```

**Comment by konstruktoid** [comment]
> Looks like you're trying to validate that the string "log/docker-bench-security.sh.log.json" is a valid json.
> I think a better approach is `cat log/docker-bench-security.log.json | jq empty`.

You are correct, `cat log/docker-bench-security.log.json | jq -r` works.

**Comment by razvanstoica89** [comment]
Did you manage to test the changes I made?

**Comment by konstruktoid** [comment]
> Did you manage to test the changes I made?

Yeah, but https://github.com/docker/docker-bench-security/pull/467#issuecomment-809282467 still exists.

**Comment by razvanstoica89** [comment]
> > Did you manage to test the changes I made?
> 
> Yeah, but [#467 (comment)](https://github.com/docker/docker-bench-security/pull/467#issuecomment-809282467) still exists.

Yes. You are right. By d0443cc I solved the problem reported in https://github.com/docker/docker-bench-security/pull/467#issuecomment-809282467 but I didn't let you know.

**Comment by konstruktoid** [comment]
```sh
~$ git clone https://github.com/docker/docker-bench-security
Cloning into 'docker-bench-security'...
[...]
~$ cd docker-bench-security/
~$ git checkout -b razvanstoica89-master master
[...]
~$ git pull git://github.com/razvanstoica89/docker-bench-security.git master
[...]
~$ git log | grep -A3 d0443c
commit d0443cc817cf17452f41510954e450a320b59c6a
Author: Razvan Stoica <razvan.stoica89@gmail.com>
Date:   Mon Mar 29 15:22:14 2021 +0300
~$ curl -sSL get.docker.com | sh
[...]
$ sudo bash ./docker-bench-security.sh
# --------------------------------------------------------------------------------------------
# Docker Bench for Security v1.3.6
#
# Docker, Inc. (c) 2015-2021
#
# Checks for dozens of common best-practices around deploying Docker containers in production.
# Inspired by the CIS Docker Benchmark v1.2.0.
# --------------------------------------------------------------------------------------------

Initializing 2021-04-08T20:42:29+00:00


Section A - Check results

[INFO] 1 - Host Configuration
[INFO] 1.1 - General Configuration
[NOTE] 1.1.1  - Ensure the container host has been Hardened (Not Scored)
[PASS] -c
[INFO]        * Using 20.10.5, verify is it up to date as deemed necessary
[INFO] 1.2 - Linux Hosts Specific Configuration
WARNING: No swap limit support
[WARN] -s
[INFO] -c
[INFO]       * Users:
[WARN] -s
[WARN] -s
[WARN] -s
[WARN] -s
[WARN] -s
[WARN] -s
[INFO] -c
[INFO]        * File not found
[INFO] -c
[INFO]         * File not found
[WARN] -s
[INFO] -c
[INFO]         * File not found

[INFO] 2 - Docker daemon configuration
[WARN] -s
[PASS] -s
[PASS] -s
[PASS] -s
[PASS] -s
[INFO] -c
[INFO]      * Docker daemon not listening on TCP
```

**Comment by razvanstoica89** [comment]
Please test it using the `sudo sh ./docker-bench-security.sh` command as recommended in the original README.md file.

**Comment by konstruktoid** [comment]
`sh` is linked to `bash` on multiple distributions.

```sh
$ sudo sh ./docker-bench-security.sh
# --------------------------------------------------------------------------------------------
# Docker Bench for Security v1.3.6
#
# Docker, Inc. (c) 2015-2021
#
# Checks for dozens of common best-practices around deploying Docker containers in production.
# Inspired by the CIS Docker Benchmark v1.2.0.
# --------------------------------------------------------------------------------------------

Initializing 2021-04-09T09:52:09+00:00


Section A - Check results

[INFO] 1 - Host Configuration
[INFO] 1.1 - General Configuration
[NOTE] 1.1.1  - Ensure the container host has been Hardened (Not Scored)
[PASS] -c
[INFO]        * Using 20.10.5, verify is it up to date as deemed necessary
[INFO] 1.2 - Linux Hosts Specific Configuration
[WARN] -s
[INFO] -c
[INFO]       * Users: 
[WARN] -s
[WARN] -s
[WARN] -s
[WARN] -s
[WARN] -s
[INFO] -c
[INFO]        * File not found
[INFO] -c
[INFO]        * File not found
[INFO] -c
[INFO]         * File not found
[WARN] -s
[INFO] -c
[INFO]         * File not found

[INFO] 2 - Docker daemon configuration
[WARN] -s
[PASS] -s
[PASS] -s
[PASS] -s
[PASS] -s
[INFO] -c
[INFO]      * Docker daemon not listening on TCP
^C
$ ls -l $(which sh)
lrwxrwxrwx. 1 root root 4 Jan 12 08:24 /usr/bin/sh -> bash
$ cat /etc/redhat-release 
CentOS Stream release 8
$ 
```

**Comment by razvanstoica89** [comment]
I think I managed to fix this bug https://github.com/docker/docker-bench-security/pull/467#issuecomment-816567569

**Comment by konstruktoid** [comment]
Looks good, but I'm not sure about those impact statements.
E.g. `[INFO] 5.31 - You should ensure that no containers mount docker.sock as a volume. Impact: None.` Mounting `docker.sock` is basically server ownership.

**Comment by razvanstoica89** [comment]
These impact statements refer to the impact of the implementation of remedial measures.
Not to the impact that the lack of implementation of remedial measures has.

Keep in mind that in very few situations or use cases, you want to mount `docker.sock` even if it is definitely not recommended.
Example: [Portainer.io] (https://documentation.portainer.io/v2.0/deploy/ceinstalldocker/)

File: tests/5_container_runtime.sh [hunk 8]
```diff
@@ -1151,7 +1151,7 @@ check_5_31() {
   local id="5.31"
   local desc="Ensure that the Docker socket is not mounted inside any containers (Scored)"
   local remediation="You should ensure that no containers mount docker.sock as a volume."
-  local remediationImpact="None."
+  local remediationImpact="If you really, really have to do this, you should use user namespaces and always ensure that the images you run with this configuration are properly audited and that you trust them in order to avoid potential breaches."
```
Comment (konstruktoid) [inline tests/5_container_runtime.sh:1154]: "None." is fine, since it's the `remediationImpact` we're talking about, why not just change "Impact" to "Remedation Impact"?

Comment (razvanstoica89) [reply tests/5_container_runtime.sh:1154]: You are right. It's better `None.` I'm rolling back my last change. Please excuse my mistake.

File: functions/output_lib.sh [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+      if [ -n "${addSpaceHeader}" ]; then
+        globalRemediation="${globalRemediation}\n"
+      fi
+      globalRemediation="${globalRemediation}\n${bldblu}[INFO]${txtrst} ${checkHeader}"
+      checkHeader=""
+      addSpaceHeader="1"
+    fi
+    globalRemediation="${globalRemediation}\n${bldblu}[INFO]${txtrst} ${id} - ${remediation}"
+    if [ -n "${remediationImpact}" ]; then
+      globalRemediation="${globalRemediation} Impact: ${remediationImpact}"
```
Comment (konstruktoid) [inline functions/output_lib.sh:187]: I meant to change this `Impact` to `Remediation Impact` as in the json file.

Comment (razvanstoica89) [reply functions/output_lib.sh]: @konstruktoid Thanks for the feedback. I made this change.

**Comment by konstruktoid** [comment]
Looks good!
And thanks for this massive, massive contribution.
